src/main/scala/dibl/PairDiagram.scala

```diff
@@ -15,143 +15,130 @@
 */
 package dibl
 
+import scala.annotation.tailrec
 import scala.util.Try
 
 case class PairDiagram private(nodes: Seq[Props],
                                links: Seq[Props])
 
 object PairDiagram {
 
-  def apply(settings: Try[Settings]): PairDiagram =
-    if (settings.isFailure)
-      PairDiagram(Seq(Props("title" -> settings.failed.get.getMessage, "bobbin" -> true)), Seq[Props]())
-    else {
-      val nodeNrs = assignNodeNrs(settings.get.absM, settings.get.nrOfPairLinks)
-      val nodes = toNodes(settings.get).toArray
-      val links = toLinks(settings.get, nodeNrs, nodes)
-      assignPairNrs(nodes, links)
-      PairDiagram(nodes, links)
+  def apply(triedSettings: Try[Settings]): PairDiagram = if (triedSettings.isFailure)
+    PairDiagram(Seq(Props("title" -> triedSettings.failed.get.getMessage, "bobbin" -> true)), Seq[Props]())
+  else {
+    val settings: Settings = triedSettings.get
+    val fringes = new Fringes(triedSettings.get.absM)
+    val sources: Seq[Cell] = fringes.newPairs.map { case (source, _) => source }
+    val plainLinks: Seq[Link] = fringes.newPairs ++ fringes.leftFootSides ++ fringes.coreLinks ++ fringes.rightFootSides
+    val linksByTarget: Map[Cell,Seq[Link]] = replaceYsWithVs(plainLinks.groupBy { case (_, target) => target })
```

Comment:
To generate a proper thread diagram out of a pair diagram, the links should be listed clock wise per target. Not 100% sure, but it looks like that `groupBy` keeps original relative positions of the links.

The diagram below illustrates an example of `plainLinks` which has problems with the order. The new pairs are red, footsides in purple and core links grey.
![naamloos](https://cloud.githubusercontent.com/assets/10553630/20639167/73126896-b3bc-11e6-87f7-1d7463273ca1.png)
For the node in the top right corner, the red link will come before the grey one, what is not clockwise. On the left we occasionally see two purple links arriving at the same target, the long one comes before the short, that is also not clock wise.

Performance wise it might be better to reuse `absM` and replace the original links with the ones altered by the `Fringes class`.